core/libraries/batch/JobHandlers/ExecuteBatchDeletion.php

```diff
@@ -147,16 +161,117 @@ public function continue_job(JobParameters $job_parameters, $batch_size = 50)
         return new JobStepResponse(
             $job_parameters,
             sprintf(
-                esc_html__('Deleted %d items.', 'event_espresso'),
+                // translators: 1: the quantity of items deleted.
+                esc_html__('Deleted %1$d items.', 'event_espresso'),
                 $units_processed
             )
         );
     }
 
+    /**
+     * @since $VID:$
+     * @param $reg_ids
+     * @param $batch_size
+     * @param JobParameters $job_parameters
+     * @return int
+     * @throws InvalidArgumentException
+     * @throws InvalidDataTypeException
+     * @throws InvalidInterfaceException
+     * @throws EE_Error
+     * @throws ReflectionException
+     */
+    protected function deleteRegistrationsAndTransactions($reg_ids, $batch_size, JobParameters $job_parameters)
+    {
+        $count_deleted = 0;
+        foreach ($reg_ids as $reg_id) {
+            $count_deleted += $this->deleteRegistrationAndTransaction($reg_id, $job_parameters);
+            if ($count_deleted >= $batch_size) {
+                return $count_deleted;
+            }
+        }
+        return $count_deleted;
+    }
+
+    /**
+     * @since $VID:$
+     * @param $reg_id
+     * @param JobParameters $job_parameters
+     * @return int
+     * @throws EE_Error
+     * @throws InvalidArgumentException
+     * @throws InvalidDataTypeException
+     * @throws InvalidInterfaceException
+     * @throws ReflectionException
+     */
+    protected function deleteRegistrationAndTransaction($reg_id, JobParameters $job_parameters)
+    {
+        // If the registration's transaction has no other valid registrations, delete it, and its dependent data too.
+        $TXN = EEM_Transaction::instance()->get_one(
+            [
+                [
+                    'Registration.REG_ID' => $reg_id
+                ]
+            ]
+        );
+        $count_deleted = 0;
+        if ($TXN instanceof EE_Transaction) {
+            $valid_regs = $TXN->count_related(
+                'Registration',
+                [
+                    [
+                        'REG_deleted' => false,
+                        'Event.status' => ['!=', EEM_Event::post_status_trashed]
+                    ]
+                ]
+            );
+            if (! $valid_regs) {
+                // delete the transaction
+                // and its dependent data
+                $transaction_as_root_of_tree = new ModelObjNode($TXN->ID(), $TXN->get_model());
+                $transaction_as_root_of_tree->visit(1000);
```

Comment:
I thought it was reasonable to assume a transaction won't have more than 1000 dependent database rows (eg registrations, extra metas, payments, line items, etc).

Reply:
eee... ya... I would hope so, but I guess you never know... does that included answers as well?

Reply:
oh yeah it does include answers... right some people sometimes have hundreds of answers (but I seriously doubt they'll make a hundred of registrations on a single transaction, right?)
If this isn't a reasonable assumption, I think we may need to try to discover what transaction we want to delete when generating the preview page instead. Plus then we wouldn't have this weird thing where the batch deletion job's "job size" can grow...

Reply:
I created https://github.com/eventespresso/event-espresso-core/pull/2247 which can help this ticket with that issue; but I also created https://github.com/eventespresso/event-espresso-core/pull/2248 based off it which is an alternative to this PR